src/core/ddsi/src/ddsi_typewrap.c

```diff
@@ -953,25 +959,27 @@ static dds_return_t xt_validate_impl (struct ddsi_domaingv *gv, const struct xt_
         return ret;
       for (uint32_t n = 0; n < t->_u.structure.members.length; n++)
       {
-        if ((ret = xt_valid_member_flags (gv, t->_u.structure.members.seq[n].flags, MEMBER_FLAG_STRUCT_MEMBER)))
+        DDS_XTypes_StructMemberFlag flags = t->_u.structure.members.seq[n].flags;
+        bool key = in_key || (flags & DDS_XTypes_IS_KEY);
```

Comment:
Would this not cause it to reject:
```
struct A { @key x; @optional y; };
struct B { @key A z; };
```
* validation starts with `B`, `in_key` = `false`
* field `z`: key flag is set, so validation of `A` starts with `in_key` = `true`
* when it gets to field `y`, `xt_valid_member_flags` is called with `in_key` = `true` and so rejects it at line 873 (`if (in_key && (flags & O))`)

In my understanding, because `A` contains some members marked `@key`, the members not marked `@key` (i.e., `y`) are not key fields. It `x` hadn't marked as key, then of course rejecting it would've been correct.

I think it needs a separate run over the structure members to check for the presence of some `@key` annotations.

Reply:
Good point, I missed that. In [88cfe84](https://github.com/eclipse-cyclonedds/cyclonedds/pull/2154/commits/88cfe84edfa1da5bb3b6d7b99ce4c8406294e44a) I've addressed this, and also take the members in the base type into account for this.